runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java

```diff
@@ -169,9 +170,19 @@ public void write(@Nonnull IToken token, @Nonnull Object data, @Nonnull CacheOpt
 
             // Do the write
             try {
-                l.getReplicationMode(token.getSequence())
-                        .getReplicationProtocol(runtime)
-                        .write(e, ld);
+                // Cache the successful write
```

Comment:
I think the comment here should have more details, sth. like atomically write to server and then cache to prevent race condition that can overwrite the undo record.

Reply:
Ditto.